### ISSUE TITLE
Youku 201

### ISSUE THREAD

**Issue by zywzyw1234567** [soimort/you-get#2057]
@rosynirvana 
新youku.py出现201报错
debug结果如下：
```
C:\Users\zywzy>you-get --debug http://v.youku.com/v_show/id_XMjc5Nzg2MzQ4MA==.html?spm=a2hzp.8253869.0.0
[DEBUG] get_content: https://ups.youku.com/ups/get.json?vid=XMjc5Nzg2MzQ4MA==&ccode=0401&client_ip=192.168.1.1&utid=4037e0d8&client_ts=1496353722
you-get: 客户端无权播放,201
```

**Comment by RogerFederer03** [comment]
Actually if I use code in 413c39e, the problem is fixed for videos like this http://v.youku.com/v_show/id_XMjc5ODQyMTQzNg==.html?spm=a2h0z.8244218.2371631.5

**Comment by RogerFederer03** [comment]
Problem should be fix by using 413c39e.

./you-get http://v.youku.com/v_show/id_XMjc5Nzg2MzQ4MA==.html?spm=a2hzp.8253869.0.0
site: 优酷 (Youku)
title: 袁视角 第48期.太平洋战争的导火索
stream:
- format: hd3
container: flv
video-profile: 1080P
size: 196.0 MiB (205480661 bytes)
m3u8_url: http://pl-ali.youku.com/playlist/m3u8?vid=XMjc5Nzg2MzQ4MA%3D%3D&type=hd3&ups_client_netip=69.123.143.155&ups_ts=1496453769&utid=DOG4EdW4qzsCAbZyXbU%2Bt7Jt&ccode=0401&psid=7a0877d79aa28985addefe1199ef2c9d&ckey=&duration=1144&expire=18000&ups_key=f6b430e3161b3c5c16953a78c9ec2f0d
# download-with: you-get --format=hd3 [URL]

Downloading 袁视角 第48期.太平洋战争的导火索.flv ...

**Comment by zywzyw1234567** [comment]
@RogerFederer03 
Thanks a lot,  make it!

**Comment by RogerFederer03** [comment]
No Problem, glad you fixed it. You should thank @rosynirvana , whose patch works.

**Comment by BenjaminGao** [comment]
@RogerFederer03 就是按照这个补丁修改还是有这个问题

**Comment by rosynirvana** [comment]
I don't think https://github.com/soimort/you-get/commit/413c39e1a40866af1231edd89ab0b3a6c9c586bb will generate a 8 bytes ```tuid```. If  someone who cannot fetch a cna cookie, is really trying to abuse it in the past 6 or 7 hours, there is a chance  youku api server woud block it. Until now,  it's not blocked according to my test. But I can never reproduce a 8 bytes tuid, who knows.

May I ask you guys not opening new issues for a unmerged pull request directly? You can reply in #2015 if you bumped into a bug **after several retries** , or comment in #2012 if it's something about the impl details.

And for @zywzyw1234567 , please close the issue if your problem fixed.

**Comment by RogerFederer03** [comment]
Got two curious questions @rosynirvana .

1. What is 8 bytes tuid? What is it used for? Any documentation or reference?

2. What is cna cookie? I tried to undstand https://github.com/soimort/you-get/pull/2015/, but it seems that I am missing  background knowledge cna. I did google search on it, but it does not help.

**Comment by rosynirvana** [comment]
It's described in #1849 and #2012 

Youku's ```ups``` interface needs a param named ```tuid```, which comes from a cookie named ```cna```. 

The server delivers ```cna``` cookie when accessing some address of ```mmstate.com```. 
When I access ```http://gm.mmstat.com/yt/ykcomment.play.commentInit?cna=```, the server always sends two ```set-cookie``` headers, first for ```cna``` and second for ```sca``` and it did work for several days.

Debug log with a 8 bytes was reported yesterday. 
I guess that under some network env the client receives only one ```set-cookie``` for ```sca``` because the value of ```sca``` cookie is 8 bytes.

**Comment by RogerFederer03** [comment]
@rosynirvana 

According to https://github.com/soimort/you-get/issues/1849#issuecomment-294179560, Also examples https://github.com/soimort/you-get/issues/2012#issuecomment-303907497. 

Correct me if I am wrong. I think we have the following procedure. We populate params in https://ups.youku.com/ups/get.json, get response, parse it and then we can start to download the videos by using our common.py. We need parameters like &utid= to be filled with valid data.

But when reading on https://github.com/soimort/you-get/issues/2012#issuecomment-303908414, and all your previous https://github.com/soimort/you-get/issues/2057#issuecomment-305970982 in this thread. cna cookie is get from mmstate.com. When you access http://gm.mmstat.com/yt/ykcomment.play.commentInit?cna=. Why cna is set to empty? Is it a valid URL?  Do we have to pass any other parameters in this url?

**Comment by rosynirvana** [comment]
It is a valid url according to RFC 3986.

Actually no param is necessary.
```
$ curl 'http://gm.mmstat.com/yt/ykcomment.play.commentInit' -I
HTTP/1.1 200 OK
Date: Sat, 03 Jun 2017 13:02:23 GMT
Content-Type: image/gif
Content-Length: 43
Connection: keep-alive
P3P: CP="NOI DSP COR CURa ADMa DEVa PSAa PSDa OUR IND UNI PUR NAV"
Set-Cookie: cna=X6G5EViZdAICAbZyXbXoQdrc; expires=Tue, 01-Jun-27 13:02:23 GMT; path=/; domain=.mmstat.com
Set-Cookie: sca=c3f38f4e; path=/; domain=.mmstat.com
Expires: Thu, 01 Jan 1970 00:00:01 GMT
Cache-Control: no-cache
Pragma: no-cache
```

Any params are OK
```
$ curl 'http://gm.mmstat.com/yt/ykcomment.play.commentInit?a=b&c=d' -I
HTTP/1.1 200 OK
Date: Sat, 03 Jun 2017 13:02:44 GMT
Content-Type: image/gif
Content-Length: 43
Connection: keep-alive
P3P: CP="NOI DSP COR CURa ADMa DEVa PSAa PSDa OUR IND UNI PUR NAV"
Set-Cookie: cna=dKG5ES8lKXQCAbZyXbV+0q67; expires=Tue, 01-Jun-27 13:02:44 GMT; path=/; domain=.mmstat.com
Set-Cookie: sca=dde59346; path=/; domain=.mmstat.com
Expires: Thu, 01 Jan 1970 00:00:01 GMT
Cache-Control: no-cache
Pragma: no-cache
```

Unless a ```cna``` cookie is sent with the request
```
$ curl 'http://gm.mmstat.com/yt/ykcomment.play.commentInit?a=b&c=d' --cookie 'cna=X6G5EViZdAICAbZyXbXoQdrc' -I
HTTP/1.1 200 OK
Date: Sat, 03 Jun 2017 13:08:24 GMT
Content-Type: image/gif
Content-Length: 43
Connection: keep-alive
P3P: CP="NOI DSP COR CURa ADMa DEVa PSAa PSDa OUR IND UNI PUR NAV"
Set-Cookie: sca=60408b5b; path=/; domain=.mmstat.com
Expires: Thu, 01 Jan 1970 00:00:01 GMT
Cache-Control: no-cache
Pragma: no-cache
```

Someone has tested thousands of times. Everytime a ```cna``` cookie was returned.

IMO it's a trivial part of ```youku.py```. The cookie won't expire in ten years and one can always load exported cookie at worst circumstance.

**Comment by RogerFederer03** [comment]
@rosynirvana  Thanks for the reply. I agree the cookie part might be trivial, but I believe it might be better if more people understand what the issue is. IMO If something goes wrong next time, it might be fixed faster.

**Comment by rosynirvana** [comment]
I agree. So everything I know about ```youku.py``` is recorded in #1849 and #2012 .

**Comment by yaojiwei520** [comment]
C:\Users\Administrator>you-get -o d:/video http://v.youku.com/v_show/id_XMjgwMjUwNjMyOA==.html?spm=a2hww.20023042.m_223495.5~5~5~5~A
you-get: [Failed] 客户端无权播放,201

**Comment by zywzyw1234567** [comment]
@yaojiwei520 
替换这个试一下，我试了可用
https://github.com/soimort/you-get/raw/413c39e1a40866af1231edd89ab0b3a6c9c586bb/src/you_get/extractors/youku.py

**Comment by BenjaminGao** [comment]
@zywzyw1234567 还是不可用，我这里替换了还是显示201

**Comment by qhzhang87** [comment]
413c39e works for most videos, but for those videos with region restrictions, when I use unblock youku (-y proxy.uku.im:443), it shows the following error:
```
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.5/bin/you-get", line 9, in <module>
    load_entry_point('you-get==0.4.750', 'console_scripts', 'you-get')()
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/you_get/__main__.py", line 92, in main
    main(**kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/you_get/common.py", line 1444, in main
    script_main('you-get', any_download, any_download_playlist, **kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/you_get/common.py", line 1357, in script_main
    download_main(download, download_playlist, args, playlist, extractor_proxy=extractor_proxy, output_dir=output_dir, merge=merge, info_only=info_only, json_output=json_output, caption=caption)
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/you_get/common.py", line 1157, in download_main
    download(url, **kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/you_get/common.py", line 1437, in any_download
    m.download(url, **kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/you_get/extractor.py", line 41, in download_by_url
    self.prepare(**kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/you_get/extractors/youku.py", line 192, in prepare
    data = youku_ups(self.vid)['data']
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/you_get/extractors/youku.py", line 40, in youku_ups
    url += '&utid=' + fetch_cna()
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/you_get/extractors/youku.py", line 27, in fetch_cna
    req = urllib.request.urlopen(url)
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/urllib/request.py", line 162, in urlopen
    return opener.open(url, data, timeout)
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/urllib/request.py", line 471, in open
    response = meth(req, response)
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/urllib/request.py", line 581, in http_response
    'http', request, response, code, msg, hdrs)
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/urllib/request.py", line 509, in error
    return self._call_chain(*args)
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/urllib/request.py", line 443, in _call_chain
    result = func(*args)
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/urllib/request.py", line 589, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 403: Forbidden
```

**Comment by rosynirvana** [comment]
@qhzhang87

Try bringing your own cookies

**Comment by qhzhang87** [comment]
@rosynirvana 
You mean adding `-c cookies.sqlite`?
I tried but it still showed the same error.

**Comment by rosynirvana** [comment]
I guess ```unblock youku``` doesn't work for ```mmstat.com```.

Now code handling loading cookies is far from robust and tends to fail silently. Perhaps you have to add code in ```common.py``` or ```youku.py``` to check whether cookies are correctly loaded.

There is a slight chance that there's no ```cna``` for ```.youku.com``` in your cookie file. But once your browser accesses ```youku.com``` this cookie will be added so I believe it won't be the problem.

Or you can patch ```fetch_cna``` and substitute it with a hard-encoded ```cna``` cookie.

**Comment by StLym** [comment]
有没有大佬解决了这个问题，换了几个youku.py也没用啊

**Comment by Cappuccin0** [comment]
Used to have same issue (i.e. errcode 201) on 0.4.750, and it can be resolved by 413c39e.

**Comment by 4ft35t** [comment]
修改过的有效，但是直接播放的话，后面的片段的地址会过期，返回 403
```bash
you-get -d  -p mpv http://v.youku.com/v_show/id_XMjg0MDc1NjcxNg\=\=
[DEBUG] get_content: https://ups.youku.com/ups/get.json?vid=XMjg0MDc1NjcxNg==&ccode=0401&client_ip=192.168.1.1&utid=P%2BvhEUo3qysCAd7RUzKldcx5&client_ts=1499135295
site:                优酷 (Youku)
title:               34 业火红莲
stream:
    - format:        hd3
      container:     flv
      video-profile: 1080P
      size:          510.5 MiB (535265042 bytes)
    # download-with: you-get --format=hd3 [URL]

audio-languages:
    - lang:          国语
      download-url:  http://v.youku.com/v_show/id_XMjg0MDc1NjcxNg==

Warning: option --sub-text-font-size was replaced with --sub-font-size and might be removed in the future.
Playing: http://61.128.155.181/697525D8CF53781C77F82340F6/0300800900594B2C854AA7011BA6A9B0E831BB-3D8A-3582-A39C-2DCA9CD8C284.flv?ccode=0401&duration=136&expire=18000&psid=eb8885750c5ba7da363b7b85cc1302d8&ups_client_netip=222.209.83.50&ups_ts=1499135296&ups_userid=&utid=P%2BvhEUo3qysCAd7RUzKldcx5&vid=XMjg0MDc1NjcxNg%3D%3D&vkey=Abc772e9294f842b36fd19514ba05e501
 (+) Video --vid=1 (h264 1920x1080 25.026fps)
 (+) Audio --aid=1 (aac 2ch 44100Hz)
AO: [coreaudio] 44100Hz stereo 2ch floatp
VO: [opengl] 1920x1080 yuv420p
AV: 00:02:16 / 00:02:16 (100%) A-V:  0.000 Dropped: 11 Cache:  0s+0KB

Playing: http://61.128.155.181/6774811DBA345823D9715C3E0A/0300800901594B2C854AA7011BA6A9B0E831BB-3D8A-3582-A39C-2DCA9CD8C284.flv?ccode=0401&duration=135&expire=18000&psid=eb8885750c5ba7da363b7b85cc1302d8&ups_client_netip=222.209.83.50&ups_ts=1499135296&ups_userid=&utid=P%2BvhEUo3qysCAd7RUzKldcx5&vid=XMjg0MDc1NjcxNg%3D%3D&vkey=Ad9f6c8abf1367f0d65bf476152af9543
 (+) Video --vid=1 (h264 1920x1080 25.026fps)
 (+) Audio --aid=1 (aac 2ch 44100Hz)
AV: 00:02:15 / 00:02:15 (100%) A-V:  0.000 Dropped: 48 Cache:  0s+0KB

Playing: http://61.128.155.181/6573DC62DD83581B6A0B64357C/0300800902594B2C854AA7011BA6A9B0E831BB-3D8A-3582-A39C-2DCA9CD8C284.flv?ccode=0401&duration=133&expire=18000&psid=eb8885750c5ba7da363b7b85cc1302d8&ups_client_netip=222.209.83.50&ups_ts=1499135296&ups_userid=&utid=P%2BvhEUo3qysCAd7RUzKldcx5&vid=XMjg0MDc1NjcxNg%3D%3D&vkey=A57eddf924c0ca30b49dc04ae4570da8e
 (+) Video --vid=1 (h264 1920x1080 25.026fps)
 (+) Audio --aid=1 (aac 2ch 44100Hz)
AV: 00:02:13 / 00:02:13 (100%) A-V:  0.000 Dropped: 4 Cache:  0s+0KB

Playing: http://61.128.155.181/6775CA93C143581B6A0B642F7D/0300800903594B2C854AA7011BA6A9B0E831BB-3D8A-3582-A39C-2DCA9CD8C284.flv?ccode=0401&duration=135&expire=18000&psid=eb8885750c5ba7da363b7b85cc1302d8&ups_client_netip=222.209.83.50&ups_ts=1499135296&ups_userid=&utid=P%2BvhEUo3qysCAd7RUzKldcx5&vid=XMjg0MDc1NjcxNg%3D%3D&vkey=Ab407c1e00a33a47095b60b5b4e1dbbb8
[ffmpeg] http: HTTP error 403 Forbidden
Failed to open http://61.128.155.181/6775CA93C143581B6A0B642F7D/0300800903594B2C854AA7011BA6A9B0E831BB-3D8A-3582-A39C-2DCA9CD8C284.flv?ccode=0401&duration=135&expire=18000&psid=eb8885750c5ba7da363b7b85cc1302d8&ups_client_netip=222.209.83.50&ups_ts=1499135296&ups_userid=&utid=P%2BvhEUo3qysCAd7RUzKldcx5&vid=XMjg0MDc1NjcxNg%3D%3D&vkey=Ab407c1e00a33a47095b60b5b4e1dbbb8.

Playing: http://61.128.155.181/6977B8C45EF32819D5284637EF/0300800904594B2C854AA7011BA6A9B0E831BB-3D8A-3582-A39C-2DCA9CD8C284.flv?ccode=0401&duration=139&expire=18000&psid=eb8885750c5ba7da363b7b85cc1302d8&ups_client_netip=222.209.83.50&ups_ts=1499135296&ups_userid=&utid=P%2BvhEUo3qysCAd7RUzKldcx5&vid=XMjg0MDc1NjcxNg%3D%3D&vkey=Abcf79e008ee7300a8d5a7a74f502be51
[ffmpeg] http: HTTP error 403 Forbidden
Failed to open http://61.128.155.181/6977B8C45EF32819D5284637EF/0300800904594B2C854AA7011BA6A9B0E831BB-3D8A-3582-A39C-2DCA9CD8C284.flv?ccode=0401&duration=139&expire=18000&psid=eb8885750c5ba7da363b7b85cc1302d8&ups_client_netip=222.209.83.50&ups_ts=1499135296&ups_userid=&utid=P%2BvhEUo3qysCAd7RUzKldcx5&vid=XMjg0MDc1NjcxNg%3D%3D&vkey=Abcf79e008ee7300a8d5a7a74f502be51.

Playing: http://61.128.155.181/6575CA9356244823527AFC3C3D/0300800905594B2C854AA7011BA6A9B0E831BB-3D8A-3582-A39C-2DCA9CD8C284.flv?ccode=0401&duration=130&expire=18000&psid=eb8885750c5ba7da363b7b85cc1302d8&ups_client_netip=222.209.83.50&ups_ts=1499135296&ups_userid=&utid=P%2BvhEUo3qysCAd7RUzKldcx5&vid=XMjg0MDc1NjcxNg%3D%3D&vkey=A6a4f8d78b41144c167e49c82581c412e
[ffmpeg] http: HTTP error 403 Forbidden
Failed to open http://61.128.155.181/6575CA9356244823527AFC3C3D/0300800905594B2C854AA7011BA6A9B0E831BB-3D8A-3582-A39C-2DCA9CD8C284.flv?ccode=0401&duration=130&expire=18000&psid=eb8885750c5ba7da363b7b85cc1302d8&ups_client_netip=222.209.83.50&ups_ts=1499135296&ups_userid=&utid=P%2BvhEUo3qysCAd7RUzKldcx5&vid=XMjg0MDc1NjcxNg%3D%3D&vkey=A6a4f8d78b41144c167e49c82581c412e.

Playing: http://61.128.155.181/6777B8C44B73381A5C1EA556D0/0300800906594B2C854AA7011BA6A9B0E831BB-3D8A-3582-A39C-2DCA9CD8C284.flv?ccode=0401&duration=138&expire=18000&psid=eb8885750c5ba7da363b7b85cc1302d8&ups_client_netip=222.209.83.50&ups_ts=1499135296&ups_userid=&utid=P%2BvhEUo3qysCAd7RUzKldcx5&vid=XMjg0MDc1NjcxNg%3D%3D&vkey=A3f3ee8b9042996af83d77e7e171e4037
[ffmpeg] http: HTTP error 403 Forbidden
Failed to open http://61.128.155.181/6777B8C44B73381A5C1EA556D0/0300800906594B2C854AA7011BA6A9B0E831BB-3D8A-3582-A39C-2DCA9CD8C284.flv?ccode=0401&duration=138&expire=18000&psid=eb8885750c5ba7da363b7b85cc1302d8&ups_client_netip=222.209.83.50&ups_ts=1499135296&ups_userid=&utid=P%2BvhEUo3qysCAd7RUzKldcx5&vid=XMjg0MDc1NjcxNg%3D%3D&vkey=A3f3ee8b9042996af83d77e7e171e4037.

Playing: http://61.128.155.181/65771409DCE3E82028B4BF21B5/0300800907594B2C854AA7011BA6A9B0E831BB-3D8A-3582-A39C-2DCA9CD8C284.flv?ccode=0401&duration=137&expire=18000&psid=eb8885750c5ba7da363b7b85cc1302d8&ups_client_netip=222.209.83.50&ups_ts=1499135296&ups_userid=&utid=P%2BvhEUo3qysCAd7RUzKldcx5&vid=XMjg0MDc1NjcxNg%3D%3D&vkey=A6e8db8b15f72f86dbed988d6661767f9
[ffmpeg] http: HTTP error 403 Forbidden
Failed to open http://61.128.155.181/65771409DCE3E82028B4BF21B5/0300800907594B2C854AA7011BA6A9B0E831BB-3D8A-3582-A39C-2DCA9CD8C284.flv?ccode=0401&duration=137&expire=18000&psid=eb8885750c5ba7da363b7b85cc1302d8&ups_client_netip=222.209.83.50&ups_ts=1499135296&ups_userid=&utid=P%2BvhEUo3qysCAd7RUzKldcx5&vid=XMjg0MDc1NjcxNg%3D%3D&vkey=A6e8db8b15f72f86dbed988d6661767f9.

Playing: http://61.128.155.181/6779023AFF63681BF101C36C9C/0300800908594B2C854AA7011BA6A9B0E831BB-3D8A-3582-A39C-2DCA9CD8C284.flv?ccode=0401&duration=125&expire=18000&psid=eb8885750c5ba7da363b7b85cc1302d8&ups_client_netip=222.209.83.50&ups_ts=1499135296&ups_userid=&utid=P%2BvhEUo3qysCAd7RUzKldcx5&vid=XMjg0MDc1NjcxNg%3D%3D&vkey=Ab3ea7bbea6f033c5329b23923ffdc760
[ffmpeg] http: HTTP error 403 Forbidden
Failed to open http://61.128.155.181/6779023AFF63681BF101C36C9C/0300800908594B2C854AA7011BA6A9B0E831BB-3D8A-3582-A39C-2DCA9CD8C284.flv?ccode=0401&duration=125&expire=18000&psid=eb8885750c5ba7da363b7b85cc1302d8&ups_client_netip=222.209.83.50&ups_ts=1499135296&ups_userid=&utid=P%2BvhEUo3qysCAd7RUzKldcx5&vid=XMjg0MDc1NjcxNg%3D%3D&vkey=Ab3ea7bbea6f033c5329b23923ffdc760.
```

**Comment by soimort** [comment]
Closed via #2015.
